game/game.py

```diff
@@ -0,0 +1,262 @@
+        for player in self.__players_queue:
+            if game_actions:
+                player.set_turn_actions(game_actions[player.name])
+
+        # start the game loop
+        self.start()
+
+    def start_menu(self) -> None:
+        try:
+            # initialize and start the pygame display manager from the main thread
+            DisplayManager(self).run()
+        finally:
+            # in case the main thread is interrupted
+            self.over.set()
+
+    def get_winner_name(self):
+        # wait for game end event
+        self.over.wait()
+
+        return self.__winner_name
+
+    def run(self) -> None:
+        self.__init_game_state()
+
+        while not self.over.is_set():
+            # start the next turn
+            self.__start_next_turn()
+
+            # handshake with players
+            self.__handle_players()
+
+        self.__end_game()
+
+    def __notify_all_players(self):
+        # release all players using their private semaphores
+        for player in self.__active_players.values():
+            player.next_turn_sem.release()
+
+    def __handle_players(self):
+        self.__notify_all_players()
+
+        # if current player isn't a remote player, end the turn
+        if not isinstance(self.__current_player, RemotePlayer) and not self.over.is_set():
+            self.__shadow_client.force_turn()
+
+        # wait for everyone to finish their turns
+        for _ in range(self.__num_players):
+            self.__turn_played_sem.acquire()
+
+    def __connect_local_player(self, player: Player) -> None:
+        game_client: GameClient = GameClient()
+        user_info: dict = game_client.login(player.name, player.password,
+                                            self.__game_name, self.__num_turns,
+                                            self.__max_players, player.is_observer)
+
+        player.add_to_game(user_info, game_client)
+        player.start()
+
+        self.__active_players[player.idx] = player
+
+    def __connect_remote_player(self, player: Player, user_info: dict) -> None:
+        player.add_to_game(user_info, self.__shadow_client)
+        player.start()
+
+        self.__active_players[player.idx] = player
+
+    def __wait_for_full_lobby(self) -> dict | None:
+        """ Return game state if the lobby is full, else None if the game was interrupted """
+        game_state: dict = self.__shadow_client.get_game_state()
+
+        while not self.over.is_set() and game_state["num_players"] != len(game_state["players"]):
+            # wait for all the players to join
+            game_state = self.__shadow_client.get_game_state()
+
+        if self.over.is_set():
+            # the game was interrupted - return
+            self.__notify_all_players()
+            return None
+
+        return game_state
+
+    def __init_game_state(self) -> None:
+        # Add the queued local players to the game
+        for player in self.__players_queue:
```

Comment:
Did really understood this concept. Is it is supposed that same game can be used by multiple local players in the same process? If yes than I would say that it is not really an expected way of how game works. Usually it is supposed that one game process is one player. This might also simplify game implementation.

Reply:
Not really a true human local player (since those don't exist here), but I envisioned it as adding multiple bot players to the same game (which is standard in many strategy games). This ended up being crucial for testing and makes the game more modular in a sense.